[WIP] r/aws_elasticsearchdomain: add support for cognito_options

aoggz opened the pull request:
Fixes #4061

Changes proposed in this pull request:

* Add support for cognito_options in Elasticsearch domain creation.
* Update documentation to reflect new functionality

Output from acceptance testing:

My test is currently failing. It looks like this is due to the 'behind-the-scenes' work that the Elasticsearch service is doing to configure Cognito to work with the cluster. It looks like the updates that the Elasticsearch service makes to the Cognito Identity Pool are being pulled in during the `refresh`, which is causing the test to fail (since tf wasn't aware of the 'initial' state).

I'm opening this PR before the this change is ready because I couldn't find an instance in the repo where a similar situation is occurring and would greatly appreciate feedback on how best to handle this!

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSElasticSearchDomain_CognitoOptions'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSElasticSearchDomain_CognitoOptions -timeout 120m
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptions
--- FAIL: TestAccAWSElasticSearchDomain_CognitoOptions (749.37s)
        testing.go:518: Step 0 error: After applying this step and refreshing, the plan was not empty:

                DIFF:

                UPDATE: aws_cognito_identity_pool.example
                  cognito_identity_providers.#:                                 "1" => "0"
                  cognito_identity_providers.188648750.client_id:               "l5vruqq3skb07fcbq2dgfbfuo" => ""
                  cognito_identity_providers.188648750.provider_name:           "cognito-idp.us-west-2.amazonaws.com/us-west-2_AhWuyNOVP" => ""
                  cognito_identity_providers.188648750.server_side_token_check: "true" => "false"

                STATE:

                aws_cognito_identity_pool.example:
                  ID = us-west-2:41618e40-c687-4206-bdc9-08f70ed00de4
                  provider = provider.aws
                  allow_unauthenticated_identities = false
                  arn = arn:aws:cognito-identity:us-west-2:143815088754:identitypool/us-west-2:41618e40-c687-4206-bdc9-08f70ed00de4
                  cognito_identity_providers.# = 1
                  cognito_identity_providers.188648750.client_id = l5vruqq3skb07fcbq2dgfbfuo
                  cognito_identity_providers.188648750.provider_name = cognito-idp.us-west-2.amazonaws.com/us-west-2_AhWuyNOVP
                  cognito_identity_providers.188648750.server_side_token_check = true
                  developer_provider_name =
                  identity_pool_name = tf_test_3433714343141331240
                  openid_connect_provider_arns.# = 0
                  saml_provider_arns.# = 0
                  supported_login_providers.% = 0
                aws_cognito_user_pool.example:
                  ID = us-west-2_AhWuyNOVP
                  provider = provider.aws
                  admin_create_user_config.# = 1
                  admin_create_user_config.0.allow_admin_create_user_only = false
                  admin_create_user_config.0.invite_message_template.# = 0
                  admin_create_user_config.0.unused_account_validity_days = 7
                  arn = arn:aws:cognito-idp:us-west-2:143815088754:userpool/us-west-2_AhWuyNOVP
                  auto_verified_attributes.# = 0
                  creation_date = 2018-07-26T10:42:17Z
                  device_configuration.# = 0
                  email_configuration.# = 0
                  endpoint = cognito-idp.us-west-2.amazonaws.com/us-west-2_AhWuyNOVP
                  lambda_config.# = 0
                  last_modified_date = 2018-07-26T10:42:17Z
                  mfa_configuration = OFF
                  name = tf-test-3433714343141331240
                  password_policy.# = 1
                  password_policy.0.minimum_length = 8
                  password_policy.0.require_lowercase = true
                  password_policy.0.require_numbers = true
                  password_policy.0.require_symbols = true
                  password_policy.0.require_uppercase = true
                  schema.# = 0
                  sms_configuration.# = 0
                  tags.% = 0
                  verification_message_template.# = 1
                  verification_message_template.0.default_email_option = CONFIRM_WITH_CODE
                  verification_message_template.0.email_message =
                  verification_message_template.0.email_message_by_link =
                  verification_message_template.0.email_subject =
                  verification_message_template.0.email_subject_by_link =
                  verification_message_template.0.sms_message =
                aws_cognito_user_pool_domain.example:
                  ID = tf-test-3433714343141331240
                  provider = provider.aws
                  aws_account_id = 143815088754
                  cloudfront_distribution_arn = dpp0gtxikpq3y.cloudfront.net
                  domain = tf-test-3433714343141331240
                  s3_bucket = aws-cognito-prod-pdx-assets
                  user_pool_id = us-west-2_AhWuyNOVP
                  version = 20180726104218

                  Dependencies:
                    aws_cognito_user_pool.example
                aws_elasticsearch_domain.example:
                  ID = arn:aws:es:us-west-2:143815088754:domain/tf-test-3433714343141331240
                  provider = provider.aws
                  advanced_options.% = 1
                  advanced_options.rest.action.multi.allow_explicit_index = true
                  arn = arn:aws:es:us-west-2:143815088754:domain/tf-test-3433714343141331240
                  cluster_config.# = 1
                  cluster_config.0.dedicated_master_count = 0
                  cluster_config.0.dedicated_master_enabled = false
                  cluster_config.0.dedicated_master_type =
                  cluster_config.0.instance_count = 1
                  cluster_config.0.instance_type = m4.large.elasticsearch
                  cluster_config.0.zone_awareness_enabled = false
                  cognito_options.# = 1
                  cognito_options.0.enabled = true
                  cognito_options.0.identity_pool_id = us-west-2:41618e40-c687-4206-bdc9-08f70ed00de4
                  cognito_options.0.role_arn = arn:aws:iam::143815088754:role/service-role/tf-test-3433714343141331240
                  cognito_options.0.user_pool_id = us-west-2_AhWuyNOVP
                  domain_id = 143815088754/tf-test-3433714343141331240
                  domain_name = tf-test-3433714343141331240
                  ebs_options.# = 1
                  ebs_options.0.ebs_enabled = true
                  ebs_options.0.iops = 0
                  ebs_options.0.volume_size = 10
                  ebs_options.0.volume_type = gp2
                  elasticsearch_version = 6.0
                  encrypt_at_rest.# = 1
                  encrypt_at_rest.0.enabled = false
                  encrypt_at_rest.0.kms_key_id =
                  endpoint = search-tf-test-3433714343141331240-inqnpmzxqaq3vopfc5dlmpa7cy.us-west-2.es.amazonaws.com
                  kibana_endpoint = search-tf-test-3433714343141331240-inqnpmzxqaq3vopfc5dlmpa7cy.us-west-2.es.amazonaws.com/_plugin/kibana/
                  tags.% = 0

                  Dependencies:
                    aws_cognito_identity_pool.example
                    aws_cognito_user_pool.example
                    aws_iam_role.example
                    aws_iam_role_policy_attachment.example
                aws_iam_role.example:
                  ID = tf-test-3433714343141331240
                  provider = provider.aws
                  arn = arn:aws:iam::143815088754:role/service-role/tf-test-3433714343141331240
                  assume_role_policy = {"Version":"2012-10-17","Statement":[{"Sid":"","Effect":"Allow","Principal":{"Service":"es.amazonaws.com"},"Action":"sts:AssumeRole"}]}
                  create_date = 2018-07-26T10:42:16Z
                  force_detach_policies = false
                  max_session_duration = 3600
                  name = tf-test-3433714343141331240
                  path = /service-role/
                  unique_id = AROAJAHBDLICLKECWAC3I

                  Dependencies:
                    data.aws_iam_policy_document.assume-role-policy
                aws_iam_role_policy_attachment.example:
                  ID = tf-test-3433714343141331240-20180726104216418000000001
                  provider = provider.aws
                  policy_arn = arn:aws:iam::aws:policy/AmazonESCognitoAccess
                  role = tf-test-3433714343141331240

                  Dependencies:
                    aws_iam_role.example
                data.aws_iam_policy_document.assume-role-policy:
                  ID = 196909887
                  provider = provider.aws
                  json = {
                  "Version": "2012-10-17",
                  "Statement": [
                    {
                      "Sid": "",
                      "Effect": "Allow",
                      "Action": "sts:AssumeRole",
                      "Principal": {
                        "Service": "es.amazonaws.com"
                      }
                    }
                  ]
                }
                  statement.# = 1
                  statement.0.actions.# = 1
                  statement.0.actions.2528466339 = sts:AssumeRole
                  statement.0.condition.# = 0
                  statement.0.effect = Allow
                  statement.0.not_actions.# = 0
                  statement.0.not_principals.# = 0
                  statement.0.not_resources.# = 0
                  statement.0.principals.# = 1
                  statement.0.principals.477085228.identifiers.# = 1
                  statement.0.principals.477085228.identifiers.1393498159 = es.amazonaws.com
                  statement.0.principals.477085228.type = Service
                  statement.0.resources.# = 0
                  statement.0.sid =
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       749.403s
GNUmakefile:20: recipe for target 'testacc' failed
make: *** [testacc] Error 1

...
```

bflad commented on aws/resource_aws_elasticsearch_domain.go:
```diff
@@ -384,6 +411,21 @@ func resourceAwsElasticSearchDomainCreate(d *schema.ResourceData, meta interface
+	if v, ok := d.GetOk("cognito_options"); ok {
+
+		options := v.([]interface{})
+		if len(options) > 1 {
```
Nitpick: This error checking is already performed by the attribute schema so it can be removed 👍 
* `MaxItems: 1` handles `len(options) > 1`
* `Required: true` on nested arguments handles `options[0] == nil`

I'd recommend going with the below to simplify this:

```go
if v, ok := d.GetOk("cognito_options"); ok && len(v.([]interface{})) > 0 {
  m := v.([]interface{})[0].(map[string]interface{})
  input.CognitoOptions = expandESCognitoOptions(m)
}
```

aoggz replied:
Oh cool! That's much cleaner!

bflad commented on aws/resource_aws_elasticsearch_domain.go:
```diff
@@ -402,6 +444,9 @@ func resourceAwsElasticSearchDomainCreate(d *schema.ResourceData, meta interface
 			if isAWSErr(err, "ValidationException", "Domain is still being deleted") {
 				return resource.RetryableError(err)
 			}
+			if isAWSErr(err, "ValidationException", "Amazon Elasticsearch must be allowed to use the passed role") {
```
❤️

bflad commented on aws/resource_aws_elasticsearch_domain.go:
```diff
@@ -634,6 +683,12 @@ func resourceAwsElasticSearchDomainUpdate(d *schema.ResourceData, meta interface
 
+	if d.HasChange("cognito_options") {
+		options := d.Get("cognito_options").([]interface{})
+		s := options[0].(map[string]interface{})
```
This line looks like it can cause a panic if `cognito_options` is removed from the configuration since `options` will have 0 elements. We should perform a length check and set the parameter to nil/empty/disabled as required, e.g.

```go
if d.HasChange("cognito_options") {
  // default to disabling (change as necessary! this could also be handled in expandESCognitoOptions)
  input.CognitoOptions = &elasticsearch.CognitoOptions{
    Enabled: aws.Bool(false),
  }
  // only enable if provided
  if v, ok := d.GetOk("cognito_options"); ok && len(v.([]interface{})) > 0 {
    m := v.([]interface{})[0].(map[string]interface{})
    input.CognitoOptions = expandESCognitoOptions(m)
  }
}
```

I'd recommend adding a second TestStep to the acceptance test that covers trying to remove the `cognito_options` to ensure we're okay here. 👍

bflad commented on aws/structure.go:
```diff
@@ -1088,6 +1088,51 @@ func flattenESClusterConfig(c *elasticsearch.ElasticsearchClusterConfig) []map[s
 	return []map[string]interface{}{m}
 }
 
+func expandESCognitoOptions(m map[string]interface{}) *elasticsearch.CognitoOptions {
```
It seems like its possible to simply some of the resource logic by accepting the `[]interface{}` from the schema here instead of the `map[string]interface{}`. Then we can handle the length/nil checks in one place.

e.g.

```go
if v, ok := d.GetOk("cognito_options"); ok {
  input.CognitoOptions = expandESCognitoOptions(v.([]interface{}))
}
```

bflad commented on aws/structure.go:
```diff
@@ -1088,6 +1088,51 @@ func flattenESClusterConfig(c *elasticsearch.ElasticsearchClusterConfig) []map[s
+	}
+
+	if aws.BoolValue(c.Enabled) {
+		if c.UserPoolId != nil {
```
Nitpick: We can remove the `nil` checks and still prevent potential panics at the same time by using the SDK provided functions like `aws.StringValue()` -- since the schema will default to `""` for `schema.TypeString` and `aws.StringValue()` returns `""` for `nil`.

```go
m["identity_pool_id"] = aws.StringValue(c.IdentityPoolId)
m["user_pool_id"] = aws.StringValue(c.UserPoolId)
m["role_arn"] = aws.StringValue(c.RoleArn)
```

bflad commented on website/docs/r/elasticsearch_domain.html.markdown:
```diff
@@ -120,6 +120,14 @@ Security Groups and Subnets referenced in these attributes must all be within th
+**cognito_options** supports the following attribute:
+
+AWS documentation: [Amazon Cognito Authentication for Kibana](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-cognito-auth.html)
+
```
This is missing the `enabled` attribute documentation 😄

aoggz replied:
:man_facepalming:

aoggz commented:
@bflad Thanks for the feedback!

I didn't find an issue already entered, but as I was about to enter one, I wondered if it makes sense to? It doesn't feel like an issue that needs fixed, since terraform is technically working as one would expect. The `lifecycle.ignore_changes` functionality looks like it would achieve the desired effect.

I'll work on implementing the changes you suggested (thanks again!) and will enter the issue if you feel like it's the way we should go.

bflad commented:
@aoggz you're probably right for now. 👍

aoggz commented:
@bflad I'm still working on incorporating your feedback, but I'm having some trouble with the tests. With the latest version, I'm seeing this issue when running my multi-step test. 
I'm still digging through it. I'm focusing on my `expand` and `flatten` functions, since it looks like there's some issue parsing the state. I'll commit as I make progress, but if this error points to an obvious issue to you, guidance would be much appreciated! :smile:  

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSElasticSearchDomain_CognitoOptions'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSElasticSearchDomain_CognitoOptions -timeout 120m
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptions
--- FAIL: TestAccAWSElasticSearchDomain_CognitoOptions (1669.84s)
        testing.go:518: Step 1 error: After applying this step, the plan was not empty:

                DIFF:

                UPDATE: aws_elasticsearch_domain.example
                  cognito_options.#: "1" => "0"

                STATE:

                aws_cognito_identity_pool.example:
                  ID = us-west-2:34a46c6d-c882-4a57-bb18-bb797f321359
                  provider = provider.aws
                  allow_unauthenticated_identities = false
                  arn = arn:aws:cognito-identity:us-west-2:143815088754:identitypool/us-west-2:34a46c6d-c882-4a57-bb18-bb797f321359
                  cognito_identity_providers.# = 1
                  cognito_identity_providers.4194570943.client_id = 4ecp6go5td2er256qv2jvolqvq
                  cognito_identity_providers.4194570943.provider_name = cognito-idp.us-west-2.amazonaws.com/us-west-2_8RLGkBim7
                  cognito_identity_providers.4194570943.server_side_token_check = true
                  developer_provider_name =
                  identity_pool_name = tf_test_8021273742209468323
                  openid_connect_provider_arns.# = 0
                  saml_provider_arns.# = 0
                  supported_login_providers.% = 0
                aws_cognito_user_pool.example:
                  ID = us-west-2_8RLGkBim7
                  provider = provider.aws
                  admin_create_user_config.# = 1
                  admin_create_user_config.0.allow_admin_create_user_only = false
                  admin_create_user_config.0.invite_message_template.# = 0
                  admin_create_user_config.0.unused_account_validity_days = 7
                  arn = arn:aws:cognito-idp:us-west-2:143815088754:userpool/us-west-2_8RLGkBim7
                  auto_verified_attributes.# = 0
                  creation_date = 2018-07-31T18:27:56Z
                  device_configuration.# = 0
                  email_configuration.# = 0
                  endpoint = cognito-idp.us-west-2.amazonaws.com/us-west-2_8RLGkBim7
                  lambda_config.# = 0
                  last_modified_date = 2018-07-31T18:27:56Z
                  mfa_configuration = OFF
                  name = tf-test-8021273742209468323
                  password_policy.# = 1
                  password_policy.0.minimum_length = 8
                  password_policy.0.require_lowercase = true
                  password_policy.0.require_numbers = true
                  password_policy.0.require_symbols = true
                  password_policy.0.require_uppercase = true
                  schema.# = 0
                  sms_configuration.# = 0
                  tags.% = 0
                  verification_message_template.# = 1
                  verification_message_template.0.default_email_option = CONFIRM_WITH_CODE
                  verification_message_template.0.email_message =
                  verification_message_template.0.email_message_by_link =
                  verification_message_template.0.email_subject =
                  verification_message_template.0.email_subject_by_link =
                  verification_message_template.0.sms_message =
                aws_cognito_user_pool_domain.example:
                  ID = tf-test-8021273742209468323
                  provider = provider.aws
                  aws_account_id = 143815088754
                  cloudfront_distribution_arn = dpp0gtxikpq3y.cloudfront.net
                  domain = tf-test-8021273742209468323
                  s3_bucket = aws-cognito-prod-pdx-assets
                  user_pool_id = us-west-2_8RLGkBim7
                  version = 20180731182757

                  Dependencies:
                    aws_cognito_user_pool.example
                aws_elasticsearch_domain.example:
                  ID = arn:aws:es:us-west-2:143815088754:domain/tf-test-8021273742209468323
                  provider = provider.aws
                  advanced_options.% = 1
                  advanced_options.rest.action.multi.allow_explicit_index = true
                  arn = arn:aws:es:us-west-2:143815088754:domain/tf-test-8021273742209468323
                  cluster_config.# = 1
                  cluster_config.0.dedicated_master_count = 0
                  cluster_config.0.dedicated_master_enabled = false
                  cluster_config.0.dedicated_master_type =
                  cluster_config.0.instance_count = 1
                  cluster_config.0.instance_type = m4.large.elasticsearch
                  cluster_config.0.zone_awareness_enabled = false
                  cognito_options.# = 1
                  cognito_options.0.enabled = false
                  cognito_options.0.identity_pool_id =
                  cognito_options.0.role_arn =
                  cognito_options.0.user_pool_id =
                  domain_id = 143815088754/tf-test-8021273742209468323
                  domain_name = tf-test-8021273742209468323
                  ebs_options.# = 1
                  ebs_options.0.ebs_enabled = true
                  ebs_options.0.iops = 0
                  ebs_options.0.volume_size = 10
                  ebs_options.0.volume_type = gp2
                  elasticsearch_version = 6.0
                  encrypt_at_rest.# = 1
                  encrypt_at_rest.0.enabled = false
                  encrypt_at_rest.0.kms_key_id =
                  endpoint = search-tf-test-8021273742209468323-4yh4rtjbyi6v3op3fupltuihhu.us-west-2.es.amazonaws.com
                  kibana_endpoint = search-tf-test-8021273742209468323-4yh4rtjbyi6v3op3fupltuihhu.us-west-2.es.amazonaws.com/_plugin/kibana/
                  tags.% = 0
                aws_iam_role.example:
                  ID = tf-test-8021273742209468323
                  provider = provider.aws
                  arn = arn:aws:iam::143815088754:role/service-role/tf-test-8021273742209468323
                  assume_role_policy = {"Version":"2012-10-17","Statement":[{"Sid":"","Effect":"Allow","Principal":{"Service":"es.amazonaws.com"},"Action":"sts:AssumeRole"}]}
                  create_date = 2018-07-31T18:27:55Z
                  force_detach_policies = false
                  max_session_duration = 3600
                  name = tf-test-8021273742209468323
                  path = /service-role/
                  unique_id = AROAI7GQH5DS4KXMMQO7Y

                  Dependencies:
                    data.aws_iam_policy_document.assume-role-policy
                aws_iam_role_policy_attachment.example:
                  ID = tf-test-8021273742209468323-20180731182755304400000001
                  provider = provider.aws
                  policy_arn = arn:aws:iam::aws:policy/AmazonESCognitoAccess
                  role = tf-test-8021273742209468323

                  Dependencies:
                    aws_iam_role.example
                data.aws_iam_policy_document.assume-role-policy:
                  ID = 196909887
                  provider = provider.aws
                  json = {
                  "Version": "2012-10-17",
                  "Statement": [
                    {
                      "Sid": "",
                      "Effect": "Allow",
                      "Action": "sts:AssumeRole",
                      "Principal": {
                        "Service": "es.amazonaws.com"
                      }
                    }
                  ]
                }
                  statement.# = 1
                  statement.0.actions.# = 1
                  statement.0.actions.2528466339 = sts:AssumeRole
                  statement.0.condition.# = 0
                  statement.0.effect = Allow
                  statement.0.not_actions.# = 0
                  statement.0.not_principals.# = 0
                  statement.0.not_resources.# = 0
                  statement.0.principals.# = 1
                  statement.0.principals.477085228.identifiers.# = 1
                  statement.0.principals.477085228.identifiers.1393498159 = es.amazonaws.com
                  statement.0.principals.477085228.type = Service
                  statement.0.resources.# = 0
                  statement.0.sid =
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       1669.869s
GNUmakefile:20: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

bflad commented:
The schema can sometimes be a little awkward with "container" attributes like `cognito_options` depending on how the API handles it. If the API continues to return `CognitoOptions` (e.g. with `Enabled` set to `false` after disabling it), then you can setup the `cognito_options` attribute to ignore the difference between the Terraform state (`cognito_options` existing with the API response) and not having it defined in the Terraform configuration (`cognito_options` missing) using a [DiffSuppressFunc](https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#diffsuppressfunc). We prefer this over using `Computed: true` on `cognito_options` since it still allows for drift detection of the child attributes.

Here's code I've used to workaround this with some other resources:

```go
DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
	if old == "1" && new == "0" {
		return true
	}
	return false
},
```

Hope this helps.

aoggz commented:
@bflad, your suggestion worked! I added another test and simplified the test methods. 


```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSElasticSearchDomain_CognitoOptions'             
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSElasticSearchDomain_CognitoOptions -timeout 120m
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove
--- PASS: TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove (1515.03s)
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptionsUpdate
--- PASS: TestAccAWSElasticSearchDomain_CognitoOptionsUpdate (1565.07s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       3080.130s
```

bflad commented:
This has been released in [version 1.30.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1300-august-02-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
